boutdata/squashoutput.py

```diff
@@ -173,30 +189,57 @@ def squashoutput(
                     "For some reason t_array has some duplicated entries in the new "
                     "and old file."
                 )
-    # Create single file for output and write data
-    with DataFile(fullpath, create=True, write=True, format=format, **kwargs) as f:
-        for varname in outputvars:
-            if not quiet:
-                print(varname, flush=True)
-
-            var = outputs[varname]
-            if append:
-                dims = outputs.dimensions[varname]
-                if "t" in dims:
-                    var = var[cropnew:, ...]
-                    varold = old[varname]
-                    var = BoutArray(numpy.append(varold, var, axis=0), var.attributes)
-
-            if singleprecision:
-                if not isinstance(var, int):
-                    var = BoutArray(numpy.float32(var), var.attributes)
-
-            f.write(varname, var)
-            # Write changes, free memory
-            f.sync()
-            var = None
-            gc.collect()
+    kwargs["format"] = format
+
+    # Create file(s) for output and write data
+    if time_split_size is None:
+        filenames = [fullpath]
+        t_slices = [slice(None)]
+    else:
+        tind = outputs.tind
+        # tind.stop + 1 - tind.start is the total number of t-indices ignoring the step.
+        # Adding tind.step - 1 and integer-dividing by tind.step converts to the total
+        # number accounting for the step.
+        nt = (tind.stop + 1 - tind.start + tind.step - 1) // tind.step
+        n_outputs = (nt + time_split_size - 1) // time_split_size
+        filenames = []
+        t_slices = []
+        for i in range(n_outputs):
+            parts = fullpath.split(".")
+            parts[-2] += str(time_split_first_label + i)
+            filename = ".".join(parts)
+            filenames.append(filename)
+            t_slices.append(slice(i * time_split_size, (i + 1) * time_split_size))
+
+    workers = SquashWorkers(
+        False if disable_parallel_write else parallel, filenames, kwargs
+    )
+
+    for varname in outputvars:
+        if not quiet:
+            print(varname, flush=True)
+
+        var = outputs[varname]
+        dims = outputs.dimensions[varname]
+        if append:
+            if "t" in dims:
+                var = var[cropnew:, ...]
+                varold = old[varname]
+                var = BoutArray(numpy.append(varold, var, axis=0), var.attributes)
+
+        if singleprecision:
+            if not isinstance(var, int):
+                var = BoutArray(numpy.float32(var), var.attributes)
+
+        if "t" in dims:
+            workers.write_data(varname, [var[t_slice] for t_slice in t_slices])
+        else:
+            workers.write_data(varname, [var])
 
+        var = None
+        gc.collect()
```

Comment:
Is this superfluous if nothing is deleted before this line?

Reply:
I think it makes sure that the array that was in `var` really gets deleted and frees the memory. I don't remember why I used `var = None` rather than `del var` - I suspect it was necessary (at the time at least), but not sure at all.